File: built-in plugins dsp/mb-reverb-ambient.cpp
/**
 * MB Ambient Verb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Ethereal ambient reverb
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_AMBIENT_H
#define MB_REVERB_AMBIENT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbAmbient : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-ambient";
    static constexpr const char* PLUGIN_NAME    = "MB Ambient Verb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float space = 0.9f;  // range [0, 1]
    float decay = 8.0f;  // range [2, 30]
    float mix = 0.5f;  // range [0, 1]
    };

    MbReverbAmbient() = default;
    ~MbReverbAmbient() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.space = std::clamp(params.space, 0f, 1f);
        params.decay = std::clamp(params.decay, 2f, 30f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Ambient Verb
        return input;
    }
};

#endif // MB_REVERB_AMBIENT_H

File: built-in plugins dsp/mb-reverb-ambient.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbAmbientPlugin: PluginDefinition = { id: 'mb-reverb-ambient', slug: 'mb-reverb-ambient', name: 'MB Ambient Verb', category: 'effect', type: 'reverb', version: '1.0.0', description: 'Ethereal ambient reverb', author: 'Max Booster', grade: 'A', parameters: [{ id: 'space', name: 'Space', type: 'float', defaultValue: 0.9, minValue: 0, maxValue: 1, automatable: true }, { id: 'decay', name: 'Decay', type: 'float', defaultValue: 8.0, minValue: 2, maxValue: 30, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { space: 0.9, decay: 8.0, mix: 0.5 } };

export default MbReverbAmbientPlugin;
